### DISCUSSION TITLE
neorv32 setup for a digilent cmod-a7-35t board

### DISCUSSION THREAD

**Discussion by biosbob** [stnolting/neorv32#604] [Q&A] [answered]
having exhausted the capacity of my icebreaker board, the cmod-a7-35t looks like a reasonable step up....

using `/neorv32_setups/vivado/arty-a7-test-setup` as my starting point, this project uses a "top-module" found in `neorv32/rtl/test_setups`....

having my own "top-module" which i've been using with icebreaker, i'm adapting this module to target the cmod-a7-35t.... 

question:  is there an equivalent to the modules found in `neorv32_setups/osflow/devices/ice40`???  the `neorv32/rtl/test_setups` seem generic, as they lack device-specific memory or PLL support....

**Comment by biosbob** [comment]
for now, i'll follow the steps outlined in chapter 13 of the user's guide....

**Comment by stnolting** [comment]
> using /neorv32_setups/vivado/arty-a7-test-setup as my starting point, this project uses a "top-module" found in neorv32/rtl/test_setups....

This file might be a little bit outdated - so use with caution 🙈

> question: is there an equivalent to the modules found in neorv32_setups/osflow/devices/ice40??? the neorv32/rtl/test_setups seem generic, as they lack device-specific memory or PLL support....

No, there are not. The NEORV32 is written in a platform/technology/vendor agnostic style. The default memory modules from `core/rtl/mem` should _infer_ block RAM for most target FPGAs (tested with Xilinx, Intel and Lattice).

PLLs and stuff like that are very platform-specific, so those cannot be inferred automatically. If you require a different clock than the one provided by your FPGA board, then you need to add this manually to your design.

**Comment by biosbob** [comment]
an update -- and some further questions....

my near-term goal is to re-use my "top-level" VHDL module within vivado....  initially, this module would be directly connected to the 12MHz clock on the CMOD board....  after that, i would use the vivado "clock wizard" to generate a 100MHz clock to drive the design....

![image](https://github.com/stnolting/neorv32/assets/57962169/da1b4e23-e976-4147-bd6e-0ae0d32837d7)

with `neorv32_em_cmod` as my top module, i was able to successfully synthesize and run on my CMOD board....  once again, the signal `clk_i` was directly connected to the 12MHz clock on the board within my `constraints.xdc` file....

referring back to the diagram, i took the following steps to create a 100MHz clock:

- configured the clock-wizard element with a 12 MHz input (which it "knew" from my CMOD) and a 100MHz output
- renamed the external pins from the clock-wizard to `clk_i` and `rstn_i` (so i won't have to change my `constraints.xdc` file
- ran synthesis/implementation/generation as before
- adjusted the `.bin` file loaded onto the CMOD SPI flash (which worked just fine in my 12MHz baseline)

after loading the 100MHz bitstream and doing a power-on-reset, the same bootloader code used in my working 12MHz baseline fails to execute correctly....  if all were going well, i'd see a few bytes coming out the UART and a few GPIOs toggling....

my first question is about the relationship between the `create_clock` command and the clocking-wizard used in my block design....  right now, `constraints.xdc` contains the following:

`set_property -dict { PACKAGE_PIN L17  IOSTANDARD LVCMOS33 } [get_ports { clk_i }];`
`##create_clock -add -name sys_clk -period 10.00 -waveform {0 5} [get_ports { clk_i }];`

do i even need the `create_clock` command (commented out here), since the clocking-wizard element of my block design has the same information (and more!!!)????

notice that the `clk_i` port is presumably referring to the *external* pin within the block design; should i be more explicit and use a different name for this pin (say, `sys_clk`)????

were i to use the `create_clock` command, don't need to connect `sys_clk` (not `clk_in`) to something on my CMOD board????

does `create_clock` actually do what the name suggests, and perhaps i do *not* need to use the clocking wizard at all????

confession:  this is really my time doing anything like this; and the documentation around vivado is quite confusing!!!!!

bottom line -- the CMOD is an attractive FPGA target, with significantly more LUTs than my iceBreaker....  all i want to do now is run at a faster speed than 12MHz 😉

**Reply by vhdlnerd** [reply]
If the input clock is 12Mhz on `clk_i`, then the clock constraint you want is something like:
`create_clock -add -name sys_clk -period 83.33 [get_ports { clk_i }];`
This creates a clock timing group named `sys_clk` (the name, `sys_clk`, is arbitrary and can be used in later timing constraints).

> does create_clock actually do what the name suggests, and perhaps i do not need to use the clocking wizard at all????

The `create_clock` command is only a timing constraint "thing" -- it does not create any FPGA logic/primitives.

Vivado is smart enough to know how the MMCM inside your clock module is configured and will create a 100Mhz clock constraint automatically (based on the clock constraint you put on `clk_i`) on the clock input to the neorv32 core.  So, Vivado will try to meet timing at 100Mhz for the neorv32 code (as you want).

One last thing: is the `reset` input on `clk_wiz_0` configured to be low-active?  The name `reset` implies its not.

I hope this helped.

**Reply by biosbob** [reply]
thanks for catching that `reset` input!!!!  but i'm still not running correctly....

for reference, here's my updated top-level design:

![image](https://github.com/stnolting/neorv32/assets/57962169/e2ffcecf-a8d1-4747-bf10-f7b3abd13249)

i've also named the external pins `sys_clk_i` and `sys_rstn_i` to avoid confusion....  finally, i now have these two lines in my `constraints.xdc` file:

`create_clock -add -name sys_clk -period 83.33 [get_ports { sys_clk_i }];`
`set_property -dict { PACKAGE_PIN L17  IOSTANDARD LVCMOS33 } [get_ports { sys_clk_i }];`

assuming `sys_clk_i` is bound to a 12MHz clock on my board (and that i've configured my clocking-wizard to generate a 100MHz output), is this correct????

are there some timing reports i can look at????

**Reply by vhdlnerd** [reply]
> assuming sys_clk_i is bound to a 12MHz clock on my board (and that I've configured my clocking-wizard to generate a 100MHz output), is this correct????

Yes, this is correct.

You said you had it working at 12MHz, correct?  If so, in the newer, faster version did you set the `CLOCK_FREQUENCY` generic on the `neorv32_top` component (in the `neorv32_em_cmod_0` block) to 100MHz instead of 12MHz?  This may be a generic on your  `neorv32_em_cmod_0` module that is passed down to `neorv32_top`.  If this is still set to 12MHz, then UART and SPI would now be running about 8 times too fast!

I do not use the Vivado GUI to build things most of the time -- I use scripts to build my designs. (You should stick with the GUI, though!) But, if I remember correctly, when you build via the GUI it generates a timing report named `*_timing_summary_routed.rpt`  (the first part of the file name will be your project name) -- the file is somewhere in the work directory.

Also, the `Project Summary - Overview` in the GUI should tell if you met timing (this example shows timing was **not** met):

![image](https://github.com/stnolting/neorv32/assets/1608856/e22a10b1-fbc6-4223-a26d-d81c69902be9)

**Reply by biosbob** [reply]
first things first -- thanks for the help @vhdlnerd 

yes, i did set my `CLOCK_FREQUENCY` generic to 100000000....

as for the timing parameters you show in the project summary, everything is "normal" on my end.... when i was fiddling around earlier, however, i did get some negative values here; the implementation run failed in these situations....

sometimes the implementation run would complete successfully, but with some critical warnings marked `TIMING-6`....  my latest run does not report these warning in the messages pane; but if i looked at the corresponding `*_timing_summary_routed.rpt` file i do see this warning:

![image](https://github.com/stnolting/neorv32/assets/57962169/5f7f14b1-f198-411e-b414-1f0957ba06f7)

the explanation of this warning can be found [here](https://docs.xilinx.com/r/2021.1-English/ug906-vivado-design-analysis/Example?tocId=fm4D05PoicImmm8kl96Y5A)....  i'll definitely need some help decoding this stuff 😉

**Reply by vhdlnerd** [reply]
If the 12MHz and 100Mhz clocks are the only clocks in the system, then the `TIMING-6` warning should not be causing it not to work, in this case. Although, finding the source of the warning still should be done.
I need to see the timing report to know more. Can you put report (or better, the whole project zipped) in Dropbox or the like so I can access it? 
I would not have time to look at it until after work (I'm in the US, Arizona).

**Reply by biosbob** [reply]
here's a [link](https://1drv.ms/u/s!Ahww3AvJMMj6nDZeyhuxpic8Ya6H?e=b8wqGI) to a `work.zip` file containing the state of my project....  note that this project directory does *not* contain actual `.vhd` source files....  but for sure, you'll find all the timing reports....

**Reply by vhdlnerd** [reply]
I see this in the timing report:
```
Clock                               Waveform(ns)         Period(ns)      Frequency(MHz)
-----                               ------------         ----------      --------------
sys_clk                             {0.000 41.666}       83.333          12.000          
  clk_out1_em_design_clk_wiz_0_0_1  {0.000 5.000}        10.000          100.000         
  clkfbout_em_design_clk_wiz_0_0_1  {0.000 41.666}       83.333          12.000          
sys_clk_i                           {0.000 41.666}       83.333          12.000          
  clk_out1_em_design_clk_wiz_0_0    {0.000 5.000}        10.000          100.000         
  clkfbout_em_design_clk_wiz_0_0    {0.000 41.666}       83.333          12.000          
```

It looks like the clock constraints are being defined twice.  However you created this project in Vivado, it created all the clock constrains you need.  Try commenting out:
`create_clock -add -name sys_clk -period 83.33 [get_ports { sys_clk_i }];`
in your .xdc file, rebuild and see if there is only one set of clocks in the table (and hopefully, the warnings go away).

**However**, this does not explain why it is not working because it did meet timing at 100MHz.

The reset is tied to pin `A18` which is button `BTN0` on your board.  But, the Cmod A7 Ref Manual shows this:
![image](https://github.com/stnolting/neorv32/assets/1608856/a207559d-cfdd-42a4-bc1e-c9d97c60150c)

This means `BTN0` would act like a active high reset not active low.  Does it work if you hold down `BTN0`?  (Of course, it means it should have worked how you had it originally.)

**Reply by biosbob** [reply]
i found the problem -- and it was on my end....  having said that, @vhdlnerd provided the key insight....

as i said originally, i was able to directly connect the 12MHz clock to `neorv32_em_cmod.clk_i` and had everything working just....  but (as @vhdlnerd pointed out above) `BTN0` is actually active-high....  in fact, the `neorv32_em_cmod.rstn_i` input really is active high; i made the adjustment internally and should have renamed the symbol for clarity....

but when an "active-high reset" is driven by the `locked` output from the clocking wizard, we're inverted from what we should be!!!!!

i've quickly verified that everything *IS* working at 24MHz (taking a baby step)....  i'll put some pieces together, clean-up the logic, and see how fast i can actually go....

we should be able to close out this discussion shortly....

**Reply by vhdlnerd** [reply]
I'm glad you got it to work!

**Comment by biosbob** [comment, marked as answer]
here's my cleaned-up block-design:

![image](https://github.com/stnolting/neorv32/assets/57962169/350e4611-b24c-4a6f-8568-613be76ae380)

i was able to implement this design at 120MHz -- though i'll stick with 100MHz for now, as i will be adding more IP going forward....    and while this design exceeds the capacity of my iceBREAKER board, the CMOD-A7 has plenty of capacity here:

![image](https://github.com/stnolting/neorv32/assets/57962169/8bcc3c15-b53a-44bd-b00b-ee0d79b4b17d)

an interesting alternative dev board which is still <$100 -- but you do have to commit yourself to the vivado tools, which are anything but "lean-and-mean"....

i've also tried synthesizing this design for OrangeCrab -- nice, because i can use Yosys....  plenty of LUTs on this board, though max clock-speed is probably in the ~72MHz range....

**Reply by stnolting** [reply]
@biosbob 

Great that your setup works now and thanks for sharing your FPGA results!
I'm curious - what kind of processor configuration do you use (enabled ISA extensions and peripherals, memory sizes)?

@vhdlnerd 

Thank you very much for your excellent help here! :+1:

**Reply by biosbob** [reply]
here's `neorv32_em_cmod.vhd`, which serves as my top-level module:

```
library ieee;
use ieee.std_logic_1164.all;
use ieee.numeric_std.all;

library neorv32;
use neorv32.neorv32_package.all;

entity neorv32_em_cmod is
    port (
        -- Clock and Reset --
        clk_i       : in  std_ulogic; -- global clock, rising edge
        rstn_i      : in  std_ulogic; -- global reset, low-active, async
        -- GPIO (available if IO_GPIO_EN = true) --
        gpio_o        : out std_ulogic_vector(6 downto 0); -- parallel output
       -- primary UART0 (available if IO_UART0_EN = true) --
        uart0_txd_o   : out std_logic; -- UART0 send data
        uart0_rxd_i   : in  std_logic  -- UART0 receive data
    );
end entity;

architecture neorv32_em_cmod_rtl of neorv32_em_cmod is

    signal gpio_o_int : std_ulogic_vector(63 downto 0);

begin

  -- The Core Of The Problem ----------------------------------------------------------------
  -- -------------------------------------------------------------------------------------------
  neorv32_top_inst: neorv32_top
    generic map (
        -- General --
        CLOCK_FREQUENCY              => 100000000,          -- clock frequency of clk_i in Hz
        INT_BOOTLOADER_EN            => true,               -- boot configuration: true = boot explicit bootloader; false = boot from int/ext (I)MEM
        -- RISC-V CPU Extensions --
        CPU_EXTENSION_RISCV_C        => true,               -- implement compressed extension?
        CPU_EXTENSION_RISCV_M        => true,               -- implement mul/div extension?
        CPU_EXTENSION_RISCV_U        => true,               -- implement user mode extension?
        CPU_EXTENSION_RISCV_Zicntr   => true,               -- implement base counters?
        CPU_EXTENSION_RISCV_Zifencei => true,               -- implement instruction stream sync.?
        -- Tuning Options --
        FAST_MUL_EN                  => true,               -- use DSPs for M extension's multiplier
        FAST_SHIFT_EN                => true,               -- use barrel shifter for shift operations
        CPU_IPB_ENTRIES              => 2,                  -- entries in instruction prefetch buffer, has to be a power of 2, min 1
        -- Internal Instruction memory --
        MEM_INT_IMEM_EN              => true,               -- implement processor-internal instruction memory
        MEM_INT_IMEM_SIZE            => 8*1024,             -- size of processor-internal instruction memory in bytes
        -- Internal Data memory --
        MEM_INT_DMEM_EN              => true,               -- implement processor-internal data memory
        MEM_INT_DMEM_SIZE            => 8*1024,             -- size of processor-internal data memory in bytes
        -- External Interrupts Controller (XIRQ) --
        XIRQ_NUM_CH                  => 1,                  -- number of external IRQ channels (0..32)
        XIRQ_TRIGGER_TYPE            => x"ffffffff",        -- trigger type: 0=level, 1=edge
        XIRQ_TRIGGER_POLARITY        => x"ffffffff",        -- trigger polarity: 0=low-level/falling-edge, 1=high-level/rising-edge
        -- Processor peripherals --
        IO_DMA_EN                    => true,               -- implement direct memory access controller (DMA)?        
        IO_GPIO_NUM                  => 8,                  -- number of GPIO input/output pairs (0..64)
        IO_GPTMR_EN                  => true,               -- implement general purpose timer (GPTMR)?
--        IO_MTIME_EN                  => true,               -- implement machine system timer (MTIME)?
        IO_UART0_EN                  => true,               -- implement primary universal asynchronous receiver/transmitter (UART0)?
        IO_XIP_EN                    => true                -- implement execute in place module (XIP)?
    )
    port map (
        -- Global control --
        clk_i       => clk_i,           -- global clock, rising edge
        rstn_i      => rstn_i,          -- global reset, low-active, async
        -- GPIO (available if IO_GPIO_NUM > 0) --
        gpio_o      => gpio_o_int,      -- parallel output
        -- primary UART0 (available if IO_UART0_EN = true) --
        uart0_txd_o => uart0_txd_o,     -- UART0 send data
        uart0_rxd_i => uart0_rxd_i      -- UART0 receive data
    );

    gpio_o <= gpio_o_int(6 downto 0);

end architecture;
```

not all that much here -- but this design already exceeded the capacity of my iceBREAKER board....

**Reply by stnolting** [reply]
Thanks! :+1:
